classes/s6rc.bbclass

```diff
@@ -159,10 +159,52 @@ umask %s s6-setuidgid %s s6-log -d3 %s %s" %
     for template in (d.getVar('S6RC_TEMPLATES', True) or "").split():
         tree = templatetree + "/" + template
         bb.utils.mkdirhier(tree)
-        valid_files = valid_files_service + [ "check-instance" ]
+        valid_files = valid_files_service + [ "check-instance", "mode" ]
         files = getVarFlagsExpand("S6RC_TEMPLATES_%s" % template)
+        log = getVarFlagsExpand('S6RC_TEMPLATES_%s_log' % template)
+        do_logger = not "no-log" in files
 
         write_verbatim(workdir, tree, template, files, valid_files, [ "run" ])
+        # Automatic log-service generation
+        if do_logger:
+            logname = template + "-log"
+            logmode = log.get("mode", "all-in-one")
+            if logmode == "per-instance":
+                # Create a logfile for each service instance
+                tree = templatetree + "/" + template + "/log"
+                bb.utils.mkdirhier(tree)
+                logdir = log.get("dir", "/var/log/" + template + "/${instance}@" + template)
+
+                files = { "run": "#!/bin/execlineb -P\n\
+envfile ../env/instance\n\
+importas instance INSTANCE\n\
+umask %s s6-setuidgid %s foreground { mkdir -p %s } s6-log %s %s" %
+                    (log.get("umask", "0037"),
+                    log.get("user", "logger"),
+                    logdir,
+                    log.get("script", "T s100000 n10"),
+                    logdir) }
+            elif logmode == "all-in-one":
+                # Combine all instances of a template into a single log file
+                valid_files = valid_files_atomic + valid_files_service + \
+                    [ "producer-for", "consumer-for" ]
+                tree = basetree + "/" + logname
+                write_type(tree, "longrun")
+
+                files = { "run": "#!/bin/execlineb -P\n\
+piperw 0 4\n\
+foreground { s6-fdholder-store -d4 /service/s6rc-fdholder/s pipe:s6rc-%s }\n\
+umask %s s6-setuidgid %s s6-log -d3 %s %s" %
+                            (template, log.get("umask", "0037"),
+                            log.get("user", "logger"),
+                            log.get("script", "T s100000 n10"),
+                            log.get("dir", "/var/log/" + template)),
+                        "bundles": "default",
+                        "notification-fd": "3",
+                        "dependencies": "mount-temp" }
+            else:
+                raise bb.parse.SkipRecipe("'%s' in S6RC_*_%s[%s] is not a valid option. Valid options are: %s" % (logmode, logname, "mode", "all-in-one, per-instance"))
```

Comment:
Let's drop the "no-log" flag for templates and interpret an empty "mode" as no automatic logging service required.